viewer/src/main/java/nl/b3p/viewer/util/FlamingoCQL.java

```diff
@@ -0,0 +1,172 @@
+        int closingBrackets = 0;
+        int openBrackets = 1;
+        int endIndex = 0;
+
+        for (int i = startIndex; i < filter.length(); i++) {
+            char c = filter.charAt(i);
+            if (c == '(') {
+                openBrackets++;
+            }
+            if (c == ')') {
+                closingBrackets++;
+            }
+            if (openBrackets == closingBrackets) {
+                endIndex = i;
+                break;
+            }
+        }
+        // Part with the APPLAYER filter, possibly with nested APPLAYER/GEOMETRY/ATTRIBUTE filters
+        String appLayerPart = filter.substring(startIndex, endIndex);
+
+        // call recursively to parse out all the nested applayer filters
+        appLayerPart = processFilter(appLayerPart, em);
+
+        // Rewrite APPLAYER filter to GEOMETRY filter, so it can be used for filtering other features
+        String geometryFilter = rewriteAppLayerFilter(appLayerPart, em);
+
+        String beginpart = filter.substring(0, begin);
+        String endpart = filter.substring(endIndex + 1);
+        String result = beginpart + geometryFilter + endpart;
+        return result;
+    }
+
+    protected static String rewriteAppLayerFilter(String applayerfilter, EntityManager em) throws CQLException {
+        int firstIndex = applayerfilter.indexOf(", ");
+        int secondIndex = applayerfilter.indexOf(",", firstIndex + 1);
+
+        String attribute = applayerfilter.substring(0, firstIndex);
+        String appLayerId = applayerfilter.substring(firstIndex + 1, secondIndex);
+        String filter = applayerfilter.substring(secondIndex + 1);
+        
+        filter = filter.trim();
+        appLayerId = appLayerId.trim();
+        Long id = Long.parseLong(appLayerId);
+        
+        String geom = getUnionedFeatures(filter, id, em);
+        String nieuwFilter = "intersects (" + attribute + ", " + geom + ")";
+        return nieuwFilter;
+    }
+
+    private static String getUnionedFeatures(String filter, Long appLayerId, EntityManager em) throws CQLException {
+        try {
+            ApplicationLayer al = em.find(ApplicationLayer.class, appLayerId);
+
+            GeoService gs = al.getService();
+            Layer l = gs.getLayer(al.getLayerName(), em);
+
+            if (l.getFeatureType() == null) {
+                throw new Exception("Layer has no feature type");
+            }
+
+            FeatureSource fs = l.getFeatureType().openGeoToolsFeatureSource();
+            GeometryFactory gf = new GeometryFactory(new PrecisionModel(), 28992);
+
+            Query q = new Query(fs.getName().toString());
+            if (filter != null && !filter.isEmpty()) {
+                Filter attributeFilter = ECQL.toFilter(filter);
+                attributeFilter = (Filter) attributeFilter.accept(new ChangeMatchCase(false), null);
+
+                q.setFilter(attributeFilter);
+            }
+
+            q.setMaxFeatures(MAX_FEATURES);
```

Comment:
not sure how big a problem this would be, but having a max amount of features could lead to unexpected results if there are many more features.
Ideally a count should be done first to see how many features would fit the filter and if necessary several requests made; though on many a WFS the count will probably not give a true answer.